src/pipewire/pipewire.cc

```diff
@@ -0,0 +1,597 @@
+    void close_audio();
+
+    void period_wait();
+    int write_audio(const void * data, int length);
+    void drain();
+
+    int get_delay();
+
+    void pause(bool pause);
+    void flush();
+
+private:
+    bool init_core();
+    bool init_stream();
+    void init_buffer();
+
+    struct pw_stream * create_stream();
+    bool connect_stream(enum spa_audio_format format);
+
+    static void on_core_event_done(void * data, uint32_t id, int seq);
+    static void on_registry_event_global(void * data, uint32_t id, uint32_t permissions,
+                                         const char * type, uint32_t version,
+                                         const struct spa_dict * props);
+    static void on_state_changed(void * data, enum pw_stream_state old,
+                                 enum pw_stream_state state, const char * error);
+    static void on_process(void * data);
+    static void on_drained(void * data);
+
+    static enum spa_audio_format to_pipewire_format(int format);
+    static void set_channel_map(struct spa_audio_info_raw * info, int channels);
+
+    struct pw_thread_loop * m_loop = nullptr;
+    struct pw_stream * m_stream = nullptr;
+    struct pw_context * m_context = nullptr;
+    struct pw_core * m_core = nullptr;
+    struct pw_registry * m_registry = nullptr;
+
+    struct spa_hook m_core_listener = {};
+    struct spa_hook m_stream_listener = {};
+    struct spa_hook m_registry_listener = {};
+
+    bool m_inited = false;
+    bool m_has_sinks = false;
+    bool m_ignore_state_change = false;
+
+    int m_aud_format = 0;
+    int m_core_init_seq = 0;
+
+    unsigned char * m_buffer = nullptr;
+    unsigned int m_buffer_at = 0;
+    unsigned int m_buffer_size = 0;
+    unsigned int m_frames = 0;
+    unsigned int m_stride = 0;
+    unsigned int m_rate = 0;
+    unsigned int m_channels = 0;
+};
+
+EXPORT PipeWireOutput aud_plugin_instance;
+
+const char PipeWireOutput::about[] =
+ N_("PipeWire Output Plugin for Audacious\n"
+    "Copyright 2022 Thomas Lange\n\n"
+    "Based on the PipeWire Output Plugin for Qmmp\n"
+    "Copyright 2021 Ilya Kotov");
+
+const char * const PipeWireOutput::defaults[] = {
+    "volume_left", "50",
+    "volume_right", "50",
+    nullptr
+};
+
+StereoVolume PipeWireOutput::get_volume()
+{
+    return {aud_get_int("pipewire", "volume_left"),
+            aud_get_int("pipewire", "volume_right")};
+}
+
+void PipeWireOutput::set_volume(StereoVolume v)
+{
+    aud_set_int("pipewire", "volume_left", v.left);
+    aud_set_int("pipewire", "volume_right", v.right);
+
+    if (!m_loop)
+        return;
+
+    float values[m_channels];
+
+    if (m_channels == 2)
+    {
+        values[0] = v.left / 100.0f;
+        values[1] = v.right / 100.0f;
+    }
+    else
+    {
+        for (unsigned int i = 0; i < m_channels; i++)
+            values[i] = aud::max(v.left, v.right) / 100.0f;
+    }
+
+    pw_thread_loop_lock(m_loop);
```

Comment:
Should the `lock` be done at the beginning of the method? Or does it not matter here? Same question for `init_stream()`.

Reply:
I think they are fine where they are :+1: